service/resource/configmap/resource.go

```diff
@@ -188,7 +189,61 @@ func (r *Resource) GetDeleteState(ctx context.Context, obj, currentState, desire
 }
 
 func (r *Resource) GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
-	return nil, nil, nil, nil
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	currentConfigMaps, err := toConfigMaps(currentState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	desiredConfigMaps, err := toConfigMaps(desiredState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+
+	var configMapsToCreate interface{}
+	{
+		configMapsToCreate, err = r.GetCreateState(ctx, obj, currentState, desiredState)
+		if err != nil {
+			return nil, nil, nil, microerror.Mask(err)
+		}
+	}
```

Comment:
This is something I thought about. We want to run the create logic before the update logic. [ProcessUpdate](https://github.com/giantswarm/operatorkit/blob/71a88e7f33f1ea1fb059e7afddfc7a49f2ce53c3/framework/framework.go#L275-L331) is already doing this. The [Framework.UpdateFunc](https://github.com/giantswarm/operatorkit/blob/71a88e7f33f1ea1fb059e7afddfc7a49f2ce53c3/framework/framework.go#L138-L140) now redirects to the [Framework.AddFunc](https://github.com/giantswarm/operatorkit/blob/71a88e7f33f1ea1fb059e7afddfc7a49f2ce53c3/framework/framework.go#L66-L97) which calls `ProcessUpdate` as well. So most of the time we create, create and update. The question is how to improve this. 

* When we keep it like this we do unnecessary work and create extra pressure on the systems. I vote against this. 
* When we remove the creation part shown in this diff and do not create within `GetUpdateState` we have to change the interface. The interface change would imply that during updates we have to rely on `ProcessCreate` as it is. This might be good enough, but maybe not. I do not know at this point. This option would also mean the interface is not that consistent since then only the deletion and update process is being managed within `GetUpdateState`. I vote against this. 
* When we do not redirect `Framwork.UpdateFunc` to `Framework.AddFunc` the creation is still done before the update process but we fix the actual duplicated creation process. `Framework.UpdateFun` would then get its own implementation like `Framework.AddFunc` and `Framework.DeleteFunc` where we then only call `ProcessUpdate`. This implies we always have to implement the create-state creation like seen above. This is probably a little bit redundant in most cases but open enough to cover other scenarios we cannot think of now. I vote for this way to go. 

The state of the art does not really hurt at this point. It is not critical to sort this out with this PR. We can solve this separately. I just wanted to raise awareness and start the discussion so that we can sort it out in the next days.

Reply:
My opinion:

* `ProcessUpdate` may and should rely on the fact that it's called after successful call to `ProcessCreate`.
* If `ProcessUpdate` duplicates the logic of `ProcessCreate` we don't need both.
* I favour correctness more than performance. If we go back to not redirect `update` to `add` we again have problems with processing partially processed resources.
* Way to improve performance is to use raw API watch instead of caching mechanism provided by `client-go` library. That way we won't do redundant `add` and `update` calls.

Reply:
> `ProcessUpdate` may and should rely on the fact that it's called after successful call to `ProcessCreate`.

Agreed. This is and always will be the case. 

> * If `ProcessUpdate` duplicates the logic of `ProcessCreate` we don't need both.

We would not "need" it, but I would really like to keep this separation, even if the logic of `ProcessCreate` is also covered in `ProcessUpdate`. 

> I favour correctness more than performance. If we go back to not redirect `update` to `add` we again have problems with processing partially processed resources.

Agreed. Nobody wants to sacrifice correctness because of performance. I just tried to draw a picture of the things I see. 

> Way to improve performance is to use raw API watch instead of caching mechanism provided by `client-go` library. That way we won't do redundant `add` and `update` calls.

The performance part is not the issue I wanted to talk about here. There is more to this than that and there are a lot of things we can do to improve it. 

---

I will provide a PR in operatorkit which we can use to talk about. Having code is easier to reason about. Thanks for your input so far though.